### PR TITLE
Integrating context and discogs typeaheads

### PR DESCRIPTION
**Pull request by hudajkhan**
Setting up a draft PR to gather feedback on approach.
Planned additional changes listed below after description of PR.

- Adds input QA components that provide context (InputLookupQAContext), allows discogs information to be retrieved (InputLookupQADiscogs). The original typeahead functionality (InputLookupQATypeahead) is also available.  
- InputLookupQA contains the main search code and then determines which component above should be used based on the presence of "subtype" in the resource template.  For example, in the _fixtures_ templates, MonographInstance.json now includes "subtype":"context" in addition to "type":"lookup".  If subtype is not present or does not map to "context" or "discogs", the default typeahead lookup is employed.
- To support Discogs, the apidoc.json was updated to enable paths which are for sources that are not provided as linked data by their original APIs (since Discogs is not provided by its source API as linked data).  The paths supporting search and show using patterns like /show/{vocab} or /search/{vocab} have been added (all the original paths such as /show/linked_data/{vocab} and /search/linked_data/{vocab} are still supported).  Enabling the addition of these paths allows for the application to make calls using QA that are both originally linked data as well as non-linked data sources.  
- Additionally, the lookupConfig.json file has Discogs added as an authority with a flag "nonldLookup" that is used by the InputQA search code to determine which path should be used to make the call.  
- The search method has been pulled out into InputQA and is passed as a prop to the other components to call.  The options and isLoading variables which are stored in InputQA component state are passed as properties as well to the other components.  Options stores the result of the search and isLoading is used as a prop on the typeahead component as well as a way to signal the interface in the InputLookupQAContext component to show the loading icon when the search button is pressed.
- The InputLookupQAContext component allows the user to input the query and click the "search" button which opens up a modal that displays the search results along with contextual information.  The order and selection of context properties has been hard-coded (for now) for two subauthorities (LCNAF persons and organizations) whereas for others all context (other than authoritative label and preferred label which should already be reflected in the title of the result) will be returned.  
- The InputLookupQADiscogs component differs from regular typeahead in that it shows additional context such as image and other info in the dropdown.  
- In addition to updating the MonographInstance.json template in the fixtures directory to include context, I've also added  Discogs and ShareVDE templates that demonstrate discogs info using the discogs typeahead and querying the QA ShareVDE content respectively.  For the latter, the results depend on what information has been made available and what version has been indexed so far, but the templates offer a way of using both context and typeahead for works and instances.  

Additional work that needs to be done (that I plan on doing soon):
Update validation method for context (since it's not a typeahead)
Reviewing errors and warnings coming back from eslint.  I've reviewed some of them but will step through these (especially the errors).
Review what tests need to be written for the new components that are not covered by the original ones for InputLookupQA. 
Review additional refactoring that may be possible, as there is code still being repeated across the different components.  The main search code is now in one main function but it will be good to see how much more can be taken out. 

I'm setting up the draft PR now to gather feedback for any changes that may need to be made and/or make sure the approach used here in general fits with the expectations for the rest of the code.

### CODE REVIEW

File: .settings/ts.eclipse.ide.core.prefs [hunk 1]
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
```
[the remark this answers is not in the document]
Comment (hudajkhan) [reply .settings/ts.eclipse.ide.core.prefs]: Thanks!

**Comment by hudajkhan** [comment]
Following changes so far: There are now two main components (instead of three) as defined above that handle the display while InputLookupQA handles delegation to these components as well as the main search.  

Per continuing discussions with @jermnelson , we will work on breaking this PR into smaller PRs.

**Comment by hudajkhan** [comment]
Smaller PRs may include:

1.  Updating the apidoc to allow for calls to sources such as Discogs (that do not live at a /linked_data address).  This PR would include updates to apidoc, lookupconfig, and a few lines added to the search code to enable the appropriate authority lookup to be used for non-linked data sources. 
   - To support Discogs, the apidoc.json was updated to enable paths which are for sources that are not provided as linked data by their original APIs (since Discogs is not provided by its source API as linked data). The paths supporting search and show using patterns like /show/{vocab} or /search/{vocab} have been added (all the original paths such as /show/linked_data/{vocab} and /search/linked_data/{vocab} are still supported). Enabling the addition of these paths allows for the application to make calls using QA that are both originally linked data as well as non-linked data sources.
- Additionally, the lookupConfig.json file has Discogs added as an authority with a flag "nonldLookup" that is used by the InputQA search code to determine which path should be used to make the call.
- Adding code to the search method to determine which action to use based on whether or not it's a linked data source (as well as the existing criteria).  

2. Updating InputLookupQA to incorporate additional context (but still in the typeahead mode) - includes changes to the component.  To support discogs context, includes addition of CSS to support display of pictures/context 

3. Breaking out InputLookupQA into a component that is responsible for search while using a separate typeahead component for display (InputLookupQATypeahead).  

4. Adding the InputLookupQAContext as a second component and enabling InputLookupQA to select this component.

**Comment by hudajkhan** [comment]
Next steps (thanks for the meeting @jermnelson @jcoyne and @justinlittman) for now are:
- Go ahead with PR 1 above.  
- Break out context rendering code (currently in InputLookupQATypeahead) into a component (call it "ContextRenderer" perhaps).  Possibly add in code for including all the context for components that is rendered in InputQALookupContext to enable display of that information sooner rather than later. 
- Based on input/feedback from Michelle, Astrid, team, catalogers, metadata folks, decide whether modal context view should be default or an option that can be configured (the code in this branch right now enables for configuration-based option i.e. adding "subtype:context" to resource template for lookup will trigger context modal display) 

If the above should be modified, please feel free to do so/leave a comment/let me know. 

Also of note:
Possible interaction changes to context modal: remove input area on the form and have only the search input on the modal be the main point of entry to prevent confusion regarding what/where to put in search term.    Context modal would also need a way to add a URI or literal (functionality that the typeahead currently employs).  

Is there a way to make the Swagger spec not care if the path for the authority includes /linked_data or not? Maybe? (I don't have the answer but worth noting that it was a question that came up during this discussion). 

I'll be setting up specific branches for the new PRs.

**Comment by jermnelson** [comment]
@hudajkhan can we remove this PR?

**Comment by hudajkhan** [comment]
Closing as context merged into typeahead and discussions for other interaction paradigms can follow later.

**Pull request closed without merging**